rest/buildlogger_routes.go

```diff
@@ -167,16 +172,141 @@ func (h *logMetaGetByTaskIDHandler) Factory() gimlet.RouteHandler {
 // Parse fetches the id from the http request.
 func (h *logMetaGetByTaskIDHandler) Parse(_ context.Context, r *http.Request) error {
 	h.id = gimlet.GetVars(r)["id"]
+	vals := r.URL.Query()
+	h.tags = vals["tags"]
 
 	return nil
 }
 
 // Run calls FindLogMetadataByTaskID and returns the merged logs.
 func (h *logMetaGetByTaskIDHandler) Run(ctx context.Context) gimlet.Responder {
-	apiLogs, err := h.sc.FindLogMetadataByTaskID(ctx, h.id)
+	apiLogs, err := h.sc.FindLogMetadataByTaskID(ctx, h.id, h.tags...)
 	if err != nil {
 		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "Error getting log metadata by task id '%s'", h.id))
 	}
 
 	return gimlet.NewJSONResponse(apiLogs)
 }
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// GET /buildlogger/test_name/{task_id}/{test_name}
+
+type logGetByTestNameHandler struct {
+	id   string
+	name string
+	tags []string
+	tr   util.TimeRange
+	sc   data.Connector
+}
+
+func makeGetLogByTestName(sc data.Connector) gimlet.RouteHandler {
+	return &logGetByTestNameHandler{
+		sc: sc,
+	}
+}
+
+// Factory returns a pointer to a new logGetByTestNameHandler.
+func (h *logGetByTestNameHandler) Factory() gimlet.RouteHandler {
+	return &logGetByTestNameHandler{
+		sc: h.sc,
+	}
+}
+
+// Parse fetches the id, name, time range, and tags from the http request.
+func (h *logGetByTestNameHandler) Parse(_ context.Context, r *http.Request) error {
+	var err error
+
+	h.id = gimlet.GetVars(r)["id"]
+	h.name = gimlet.GetVars(r)["name"]
+	vals := r.URL.Query()
+	h.tags = vals["tags"]
+	if vals.Get(logStartAt) != "" || vals.Get(logEndAt) != "" {
+		h.tr, err = parseTimeRange(vals, logStartAt, logEndAt)
+	}
+
+	return err
+}
+
+// Run calls FindLogsByTestName and returns the merged logs.
+func (h *logGetByTestNameHandler) Run(ctx context.Context) gimlet.Responder {
+	its := []dbModel.LogIterator{}
+
+	if h.tr.IsZero() {
+		testLogs, err := h.sc.FindLogMetadataByTestName(ctx, h.id, h.name, h.tags...)
+		if err != nil {
+			return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "Error getting log metadata by test name '%s'", h.name))
+		}
+
+		for _, log := range testLogs {
+			if h.tr.StartAt.After(time.Time(log.CreatedAt)) || h.tr.StartAt.IsZero() {
+				h.tr.StartAt = time.Time(log.CreatedAt)
+			}
+			if h.tr.EndAt.Before(time.Time(log.CompletedAt)) {
+				h.tr.EndAt = time.Time(log.CompletedAt)
+			}
+		}
+	}
+
+	it, err := h.sc.FindLogsByTestName(ctx, h.id, h.name, h.tr, h.tags...)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "Error getting logs by test name '%s'", h.name))
+	}
+	its = append(its, it)
+
+	it, err = h.sc.FindLogsByTestName(ctx, h.id, "", h.tr, h.tags...)
+	if err == nil {
+		its = append(its, it)
+	} else if !strings.Contains(err.Error(), http.StatusText(http.StatusNotFound)) {
```

Comment:
would errors.Cause()? 

in evergreen we check for gimlet.ErrorResponse typed errors. the http status text and strings.Contains seems a bit weird. What about other errors?

Reply:
This is a weird case because I want to check for a not found error (meaning the database query returned nothing) and not do anything if that happens, otherwise return an error. Since it is looking for logs with an an empty `TestName` field, I do not necessarily want to error if there are none found (because there are already test level logs found). Unfortunately the `gimlet.ErrorResponse` does implement the `Cause` interface so I can't use `errors.Cause()`.

Reply:
I think casting is more clear in this case.